### PR TITLE
Fix race condition between CTS.Cancel and Dispose

### PR DESCRIPTION
**Pull request by stephentoub**
In general, CancellationTokenSource.Dispose is not meant to be used concurrently with other operations on the instance: Dispose should only be used when the creator of the CTS is done with it.  However, there's one common situation where recommended usage runs afoul of this.  A typical usage of a LinkedCancellationTokenSource (as created by CancellationTokenSource.CreateLinkedTokenSource) is as follows:
```C#
private CancellationToken _internalCancellation;

public void SomeMethod(CancellationToken externalCancellation)
{
    using (var cts = CancellationTokenSource.CreateLinkedTokenSource(externalCancellation, _internalCancellation))
    {
        DoWork(cts.Token);
    }
}
```
where an LCTS is used to combine some external source of cancellation with some internal one.  It's important that the LCTS then be disposed, as otherwise it'll end up leaking state into constituent CTs, and in particular in the case of the externally provided one, that could be a very long lived token.  However, it's perfectly fine for the external token to have cancellation requested at any time during this method, which means that even though Dispose and Cancel weren't designed to be safe to execute concurrently, and even though we recommend against it, it's actually possible with recommended usage.

It's a very dificult race to reproduce, but when it does, the typical visible result is an ObjectDisposedException.  State corruption could also result.

This commit addresses that in two ways:
1. Rather than having the LCTS register a delegate with each CT that calls Cancel, it instead has its delegate call NotifyCancellation.  This is the same method Cancel calls, but without the upfront disposal check.  That eliminates the easily visible result of a race that would cause an ObjectDisposedException.
2. With all of the tweaking that's been done to Dispose since it was originally written, it's actually very close to being safe to use concurrently with Cancel; this takes it the rest of the way (though we still don't want to document it as such, it's really just to make LCTS work correctly, and as such Cancel does still have a disposed check at the very beginning).  The only state Dispose mutates that could be problematic is a Timer and a ManualResetEvent.  Timer's Dispose is already thread-safe.  MRE's isn't, so if it's been allocated (which is relatively rare), I add a single Interlocked.Exchange to null it out, and then do a volatile read on a state field to see whether cancellation is currently in progress, only Dispose'ing of the MRE if it is, and leaving it for finalization otherwise.  On the Cancel side, we just add some null checks to ensure we're not attempting to Dispose of null'd out fields, again with volatile reads where necessary.

Fixes https://github.com/dotnet/coreclr/issues/4167
cc: @ericeil, @kouvel, @AlfredoMS, @lukazoid

### CODE REVIEW

**Comment by stephentoub** [comment]
The OSX and Ubuntu Checked failures appear to be preexisting failures in StringMarshaling tests:
```
11:00:25 FAILED   - Interop/StringMarshalling/LPSTR/LPSTRTest/LPSTRTest.sh
11:00:25                BEGIN EXECUTION
11:00:25                /Users/dotnet-bot/j/workspace/dotnet_coreclr/master/checked_osx_tst_prtest/bin/tests/Windows_NT.x64.Checked/Tests/coreoverlay/corerun LPSTRTest.exe
11:00:25                 === Fail: Method PInvokeDef.Marshal_InOut[Managed Side],The Return string is wrong
11:00:25                	Expected:a
11:00:25                	Actual:
11:00:25                 === Fail: Method PInvokeDef.Marshal_Out[Managed Side],The Return string is wrong
11:00:25                	Expected:a
11:00:25                	Actual:
11:00:25                 === Fail: Method PInvokeDef.MarshalPointer_InOut[Managed Side],The Return string is wrong
11:00:25                	Expected:a
11:00:25                	Actual:
11:00:25                 === Fail: Method PInvokeDef.MarshalPointer_InOut[Managed Side],The Passed string is wrong
11:00:25                	Expected: Native
11:00:25                	Actual: Nativ
11:00:25                 === Fail: Method PInvokeDef.MarshalPointer_Out[Managed Side],The Return string is wrong
11:00:25                	Expected:a
11:00:25                	Actual:
11:00:25                 === Fail: Method PInvokeDef.MarshalPointer_Out[Managed Side],The Passed string is wrong
11:00:25                	Expected: Native
11:00:25                	Actual: Nativ
11:00:25                 === Fail: Method PInvokeDef.MarshalStrB_InOut[Managed Side],The Return string is wrong
11:00:25                	Expected:a
11:00:25                	Actual:
11:00:25                 === Fail: Method PInvokeDef.MarshalStrB_InOut[Managed Side],The Passed string is wrong
11:00:25                	Expected: Native
11:00:25                	Actual: Nativ
11:00:25                 === Fail: Method PInvokeDef.MarshalStrB_Out[Managed Side],The Return string is wrong
11:00:25                	Expected:a
11:00:25                	Actual:
11:00:25                 === Fail: Method PInvokeDef.MarshalStrB_Out[Managed Side],The Passed string is wrong
11:00:25                	Expected: Native
11:00:25                	Actual: Nativ
11:00:25                 === Fail: Method Del_Marshal_InOut[Managed Side],The Return string is wrong
11:00:25                	Expected:a
11:00:25                	Actual:
11:00:25                 === Fail: Method Del_MarshalPointer_Out[Managed Side],The Return string is wrong
11:00:25                	Expected:a
11:00:25                	Actual:
11:00:25                 === Fail: Method Del_MarshalPointer_Out[Managed Side],The Passed string is wrong
11:00:25                	Expected: Native
11:00:25                	Actual: Nativ
11:00:25                FAIL - 13 failure(s) occurred
11:00:25                Expected: 100
11:00:25                Actual: 101
11:00:25                END EXECUTION - FAILED
```

File: src/mscorlib/src/System/Threading/CancellationTokenSource.cs [hunk 1]
```diff
@@ -556,18 +556,33 @@ protected virtual void Dispose(bool disposing)
...
+                // we only want to do so if it's not being used by Cancel concurrently.  First, we
+                // interlocked exchange it to be null, and then we check whether cancellation is currently
+                // in progress.  NotifyCancellation will only try to set the event if it exists after it's
+                // transitioned to and while it's in the NOTIFYING state.
                 if (m_kernelEvent != null)
                 {
-                    m_kernelEvent.Close(); // the critical cleanup to release an OS handle
-                    m_kernelEvent = null; // free for GC.
+                    ManualResetEvent mre = Interlocked.Exchange(ref m_kernelEvent, null);
+                    if (mre != null && m_state != NOTIFYING) // volatile read of m_state, after read of m_kernelEvent, to synchronize with Cancel
```
Comment (ericeil) [inline src/mscorlib/src/System/Threading/CancellationTokenSource.cs]: This comment is a little misleading perhaps.  The volatile read actually does nothing to preserve the desired ordering here (it's a "downward" fence only).  Instead, ordering is ensured by the full fence in `Interlocked.Exchange`.

**Comment by ericeil** [comment]
Looks good, aside from the one comment.

File: src/mscorlib/src/System/Threading/CancellationTokenSource.cs (hunk 1 above)
Comment (stephentoub) [inline src/mscorlib/src/System/Threading/CancellationTokenSource.cs]: Removed the comment.

**Comment by omariom** [comment]
@stephentoub 
Does it make all CTSs safe to dispose or only linked ones?

**Comment by stephentoub** [comment]
> Does it make all CTSs safe to dispose or only linked ones?

See my PR comments:
> With all of the tweaking that's been done to Dispose since it was originally written, it's actually very close to being safe to use concurrently with Cancel; this takes it the rest of the way (though we still don't want to document it as such, it's really just to make LCTS work correctly, and as such Cancel does still have a disposed check at the very beginning)

**Comment by GSPP** [comment]
Any reason not to document this? Seems useful behavior. @stephentoub

**Comment by stephentoub** [comment]
> Any reason not to document this? Seems useful behavior.

From my perspective:
a) It's only been done in .NET Core, not in .NET Framework.  Ideally it'd be addressed there as well, but...
b) I'm not 100% sure it addresses all cases.  I think it does, but it's possible something's still lurking.
c) We don't want to promote using instances after they've been Dispose'd; the point of Dispose is saying "I'm done with this now, you can clean up after me as I won't be using it again".  This particular case was special because the design of the types forced you into a situation where Dispose was encouraged to be used concurrently with someone else effectively calling Cancel.
d) Dispose is still not safe to be used with other operations on the CTS, and I think it likely that if folks start to see Dispose being used concurrently with Cancel, they'll assume other concurrent usage is safe.

**Pull request merged**